### ISSUE TITLE
Context menu doesn't respect screen bounds

### ISSUE THREAD

**Issue by ajitid** [microsoft/wslg#641]
# Environment

```none
Windows build number: 10.0.22000.0 
Your Distribution version: Ubuntu 21.04
Your WSLg version: 1.0.26
```

# Steps to reproduce

Open a GUI app which supports opening context menu and right click on it. (I've checked with Nautilus and Sublime Merge.)

#  Expected behavior

Context menu should appear inside the screen. 

# Actual behavior

Context menu are not trying to position themselves inside the screen and are always opened to the right of the cursor. Screenshots are of right edge of the screen:

![image](https://user-images.githubusercontent.com/26769888/150180585-a4f8e5bf-9541-4802-9b68-2ae3f192e0b4.png)

_Nautilus_

![image](https://user-images.githubusercontent.com/26769888/150180634-40719866-3c29-4aa3-8cf2-acc163bfd144.png)

_Sublime Merge_

**Comment by hideyukn88** [comment]
@ajitid, thanks for reporting the issue, 1.0.26 is old version, would you please update to the latest WSLg (by wsl --update) and check again, below is taken from the latest version, thanks!

<img width="491" alt="image" src="https://user-images.githubusercontent.com/13857962/150214398-d452b2fb-cb89-4f04-a72c-36410130db24.png">

**Comment by ajitid** [comment]
`wsl --update` tells me that I don't have any update.

```
Checking for updates...
No updates are available.
Kernel version: 5.10.60.1
```

I might have written incorrect version there, here's is the screenshot from Apps and Features just in case:

![image](https://user-images.githubusercontent.com/26769888/150218371-d7443cef-5b1e-4881-adb7-c797265aca43.png)

Here's a GIF that might explain the issue better:-

![Animation](https://user-images.githubusercontent.com/26769888/150219481-a518bb71-13a0-4b0a-bf02-63a1a9e7fb0e.gif)

As my windows are near to right edge of the screen, when I right click on Firefox (Windows) you can see it tries to put context menu inside screen. But when I do the same in Nautilus (Ubuntu) you could see it doesn't includes screen bounds in its calculation and hence opens context menu as usual. This results in context menu going partially off the screen.

**Comment by PeterNjeim** [comment]
@ajitid He's referring to the beta releases. You can update to the latest beta (1.0.28) by downloading and running the MSI for your architecture (probably x64) from the [releases page](https://github.com/microsoft/wslg/releases). For me, my context menus do not cross the monitor's border on 1.0.28, tested in Nautilus and Microsoft Edge. I didn't test on 1.0.26 though so I don't know if the update fixed it or if it was always working.

**Comment by ajitid** [comment]
Thanks @PeterNjeim. I've udpated to latest beta. (I tried upgrading to latest WSL first but running the `Add-AppxPackage` first, but it didn't do anything for me.)

![image](https://user-images.githubusercontent.com/26769888/150283662-2bcdf456-68f8-4fd6-b1dc-3c90338f0ed9.png)

Unfortunately, I'm still facing the same issue.

**Comment by PeterNjeim** [comment]
@ajitid For me it also doesn't show the updated versions in the settings app, but the updated versions do show up in WSL itself, by running `cat /mnt/wslg/versions.txt` and `uname -a`. Here's mine: `WSLg ( x86_64 ): 1.0.30+Branch.main.Sha.18884f996f5e851af4842010b434db17d3caa7aa` and `Linux Peter-PC 5.10.81.1-microsoft-standard-WSL2`. As you can see, these are the beta versions from the WSL msixbundle.

Regardless, it seems that this bug is actually specific to your system, not WSL itself as it works fine for me. I don't know what could be causing it though. I'm using an AMD Vega iGPU (on my R5 4650G) if that means anything.

**Comment by ajitid** [comment]
> For me it also doesn't show the updated versions in the settings app, but the updated versions do show up in WSL itself, by running cat /mnt/wslg/versions.txt and uname -a. 

You are right, I can see the updated version there. 

I'll keep the issue open in case others also encounter this. Again, thanks for your help Peter!

**Comment by Bocom** [comment]
I have the same issue and upgrading to WSLg 1.3.0 as described in https://github.com/microsoft/wslg/issues/641#issuecomment-1017059179 doesn't help.

In my case it's Sublime Text's context menu and language menu (or any menu at the bottom of the screen really) that don't respect screen boundaries.

I hesitated opening an issue about this here because I'm not sure if it's actually a WSLg issue or a Sublime Text issue (it also happens in Sublime Merge for me, and I don't have any other Linux GUI application to test with installed right now).

I guess I should open an issue there (if there isn't one already).

**Comment by PeterNjeim** [comment]
@ajitid @Bocom 

I think I figured out the issue! I'm using zsh for my shell, but I setup this thing that lets me use systemd on WSL, but that was back when I was using bash, so I always type `bash` after starting a WSL shell so that I can get into "bash mode" while still having the zsh features (autocomplete mainly). Turns out, when I run nautilus within a regular zsh shell, the context menu will clip off the screen, but when I run it within bash, it'll properly move the context menu within the screen bounds. When I try to run microsoft edge within zsh, it'll just open bash for me, and I'll have to type microsoft-edge a second time to open it for real.

Another issue I've seen with zsh: I can't even right click the title bar in nautilus. I can right click the title bar in bash though.

Also, zsh seems to have a nice, rounded border with no padding: 
![image](https://user-images.githubusercontent.com/26418974/151717392-23e334a5-2c3f-4339-9dad-237ea1b21d0b.png)
whlie bash seems to have a white padding: 
![image](https://user-images.githubusercontent.com/26418974/151717398-b8dd5644-5fed-46a5-aded-8fe5c0724085.png)

Maybe WSLg isn't properly initializing the window frame on zsh.

Are you using something other than bash by any chance?

**Comment by ajitid** [comment]
hey @Bocom, if you're creating an issue on Sublime Merge/Text repo then do please mention this issue. This would help us to track it on 

@PeterNjeim I'm using fish shell, though I could confirm I always see the first screenshot you sent (with zsh) in all three — Fish, Bash and opening Files (Nautilus) using start menu.

**Comment by PeterNjeim** [comment]
Please note that sublime-merge has no issue for me, installed using snap and on bash.

**Comment by PeterNjeim** [comment]
Alright some new information: running nautilus as root in zsh changes the border to the white square bash border, and works properly. Can you test that yourself?

**Comment by ajitid** [comment]
Yep, can confirm I can see borders when running using `sudo su` and context menu remain in screen bounds.

**Comment by PeterNjeim** [comment]
Some more details I've discovered: the sudo/bash mode has a very large margin when doing a window screenshot of the nautilus window, while the zsh version has a smaller margin. The font is different, the context menu is different, even what buttons are available is different between the two modes (not even minimize and maximize buttons in zsh mode). The bash right-click menu has no margin around it and has no rounded corners, while the zsh one has rounded corners and a thin margin when doing a window screenshot of it.

sudo/bash:
![image](https://user-images.githubusercontent.com/26418974/151740211-e6322040-d155-49a9-8612-0131e9b46945.png)
![image](https://user-images.githubusercontent.com/26418974/151740213-e63d2043-2c2e-4384-b9cf-d637a22f1344.png)

zsh:
![image](https://user-images.githubusercontent.com/26418974/151740231-542dabf3-4dbd-4e6e-8fde-fa2c1c7b5364.png)
![image](https://user-images.githubusercontent.com/26418974/151740233-cebedbad-f059-49ac-9d2f-03be16f79907.png)

**Comment by PeterNjeim** [comment]
Even more information (possibly the answer): when using `xlsclients`, the sudo/bash mode properly shows nautilus:
```
Peter-PC  org.gnome.Nautilus
```
while using the zsh mode, xlsclients does not recognize nautilus.

This means sudo/bash mode is running nautilus in XWayland mode, while zsh is running it in Wayland mode.

**Comment by PeterNjeim** [comment]
@ajitid @Bocom The solution is to set the environment variable `GDK_BACKEND` to `x11`. Tested and working in zsh using `export GDK_BACKEND=x11`. Please try and see if it works.

**Comment by PeterNjeim** [comment]
I don't think this issue should be closed though, because it is indiciative of an issue with Wayland in WSLg, and should still be resolved.

**Comment by ajitid** [comment]
>  using export GDK_BACKEND=x11. Please try and see if it works.

Can confirm that it works. Yes, I'll keep the issue open as it still is a workaround for the time being

**Comment by ajitid** [comment]
@PeterNjeim I think you're exporting `DISPLAY` variable in your `.bashrc` (or in any other bash config file) which is causing "cannot find display: :0" error.

**Comment by PeterNjeim** [comment]
By choosing not to export DISPLAY variable, WSLg will no longer function, as every application will simply say "cannot open display:" (without the ":X0" this time).

**Comment by PeterNjeim** [comment]
zsh gets the same error when I set GDK_BACKEND to something that isn't `x11` or `wayland`, meaning that my bash shell seems not to be able to utilize/recognize wayland.

**Comment by ajitid** [comment]
> By choosing not to export DISPLAY variable, WSLg will no longer function, as every application will simply say "cannot open display:" (without the ":0" this time) (whether in x11 or wayland).

I meant that you maybe are _manually_ exporting DISPLAY variable in bashrc as apps run in the same way in bash as they do in my fish shell.

**Comment by PeterNjeim** [comment]
I'm not exporting it in bashrc, but I do manually export it when it's unset. Regardless, manually exporting something doesn't change anything, it's just an environment variable. Something about my bash installation is not allowing me to use wayland. I notice that when GDK_BACKEND is set to nothing in zsh (after a shell restart), wayland will be chosen, while if I do the same for bash, x11 will be chosen. In your case, wayland is chosen for both shells, meaning wayland isn't recognized by my bash shell for some reason. Note that the DISPLAY variable it also set in my zsh shell.

**Comment by Bocom** [comment]
> The solution is to set the environment variable `GDK_BACKEND` to `x11`. Tested and working in zsh using `export GDK_BACKEND=x11`. Please try and see if it works.
> 
> To revert back, set it to `wayland` instead. Note that doing this in bash will make bash fail with "cannot find display: :0" error.

Can confirm that this fixed it for me. The changes to the window chrome are a bit of a bummer, but at least it works properly.

I guess this means that there's some Wayland weirdness, then? I'm not super familiar with GDK and X11/Wayland programming, so I can only go on the fact that the X11 backend works fine. That's good information though.

**Comment by robertoalvesneto** [comment]
set `export GDK_BACKEND=wayland` in `.zshrc`  solved my interface bug and now i can use themes.
After set environment variable and run `source .zshrc` i have `display :0` problemn. But, after reboot wsl everything worked fine, both in zsh and bash.

**Comment by robertoalvesneto** [comment]
I did some tests and found that:
- if i only set `GDK_BACKEND=x11`, `GDK_BACKEND=wayland` or if I don't set anything:
    - interface boarder and themes :white_check_mark:
    - my keyboard type ptBr abnt2 :x:
- If i only set `WAYLAND_DISPLAY="wayland-1"` in my `.zshrc`
    - interface boarder and themes :x:
    - my keyboard type ptBr abnt2 :white_check_mark:
- if i set both:
    - `DISPLAY :0 ERROR`

Try export DISPLAY not work, the DISPLAY when every work fine is:
``` bash
 ❯ echo $DISPLAY
:0
```
When i get `DISPLAY ERROR` i try set many DISPLAYS, anyone work:
``` bash
#export DISPLAY=:0
#export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0
#export DISPLAY="$(grep nameserver /etc/resolv.conf | sed 's/nameserver //'):0.0"
#export DISPLAY=$(ip route|awk '/^default/{print $3}'):0.0
```
With them it would either give display error or the terminal would crash

**Comment by PeterNjeim** [comment]
Hello Roberto, are you sure you're replying to the right issue? Your comments don't seem to be related to the right-click menu appearing out of the screen bounds

**Comment by robertoalvesneto** [comment]
Hi, Peter, i have the same behavior you report in https://github.com/microsoft/wslg/issues/641#issuecomment-1025228843 and `DISPLAY :0` with the wayland, with the exception of the `menu appearing out of the screen bounds`, this was solved with the update.

**Comment by sebastianiorga** [comment]
Still seeing the bug on versions below: 

```
PS C:\Users\cseba> wsl --version
WSL version: 0.56.1.0
Kernel version: 5.10.93.2
WSLg version: 1.0.30
MSRDC version: 1.2.2924
Direct3D version: 1.601.0
Windows version: 10.0.22000.556
```

Can confirm that adding `GDK_BACKEND=x11` fixes the context menu bug but makes the overall look uglier(big boxes around the windows, different cursor).

**Comment by rowleya** [comment]
I can also confirm that ```GDK_BACKEND=x11``` certainly helps, though in eclipse it is still not perfect; sometimes the menus still drop off the bottom of the screen for some reason.  In particular, sub-menus suffer from this more.  I also noted that it works worse when using ```GDK_DPI_SCALE=1.1```, which I need to make the UI readable in GDK mode.  This is now on the latest version of WSL and WSLg:

```
WSL version: 0.65.3.0
Kernel version: 5.15.57.1
WSLg version: 1.0.41
MSRDC version: 1.2.3213
Direct3D version: 1.601.0
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22000.856
```

**Comment by rowleya** [comment]
Here is a screenshot:
![image](https://user-images.githubusercontent.com/7584931/185064146-1415e807-0c6d-4914-a99b-9bbd6c340a8c.png)

I think the issue here is that the taskbar isn't being respected.  I believe that the bottom of the menu roughly aligns with the bottom of the screen.  Indeed, setting the taskbar to autohide and then using a different monitor, I can see the bottom of the menu and it attaches to the bottom of the screen.

**Comment by topiga** [comment]
Hi ! I have a similar issue on Qt Creator. It's a bit random. Sometimes the context menu is at the right place, but often it's way to up. I encounter this issue on bash and zsh, with or without `GDK_BACKEND=x11`/`GDK_BACKEND=wayland`. It is **really** annoying. I'm using the latest version of wsl, with the preview from the microsoft store.

**Comment by speller** [comment]
I'm having the same issue with JetBraind IDEs. Taskbar size is not respected and some menu items are hidden under it. At the same time, maximization works as expected and windows are getting the correct size accounting the taskbar.

<img width="590" alt="image" src="https://user-images.githubusercontent.com/1829515/193737798-1c8d058b-cc1c-42f2-9d00-ed8424fa91c8.png">

**Comment by rowleya** [comment]
Interesting experiment; I just tried with eclipse not having the window officially "maximised" i.e. setting it to "normal" size, and then extending the window to the size of the screen instead.  This appears to result in the menu appearing *over* the taskbar rather than under it, which is a big improvement!  This is still what I would consider a workaround, but at least it appears to help...

**Comment by houdini68** [comment]
Hi,

Do you have any updates regarding this problem and a possible fix?
I’m using Eclipse under WSL, and I’m experiencing the same issue. It’s extremely frustrating.

Thank you for your help.

**Comment by Franz-Budon-Bapt** [comment]
Same issue in 2026

**Comment by rowleya** [comment]
My workaround still works for me, but still means I can't maximise the Window, and need to use "X" mode rather than Wayland, where it doesn't work at all.

**Comment by houdini68** [comment]
I had the problem yesterday because I activated WSL in Wayland mode...
Normally I use the X11 mode of WSL with an external X11 server (because of
the IBM ACE toolkit).
When starting IntelliJ, the menus for the tool window (Junie) didn't appear
entirely (our problem).
*After a long discussion with Gemini we reached the following conclusion:*

   - I switched back to WSL in X11 mode.
   - (To be honest, since I use the IBM ACE toolkit, which is an Eclipse
   application X11 mode is preferable).
   - In IntelliJ, it is possible to activate the X11 mode.
   - Intellij is displayed rather smoothly in mode X11 (I didn't notice a
   slow GUI).

My 2 cents

Have fun

Rudy

**Comment by houdini1968** [comment]
> I had the problem yesterday because I activated WSL in Wayland mode...
> Normally I use the X11 mode of WSL with an external X11 server (because of
> the IBM ACE toolkit).
> When starting IntelliJ, the menus for the tool window (Junie) didn't appear
> entirely (our problem).
> *After a long discussion with Gemini we reached the following conclusion:*
> 
>    - I switched back to WSL in X11 mode.
>    - (To be honest, since I use the IBM ACE toolkit, which is an Eclipse
>    application X11 mode is preferable).
>    - In IntelliJ, it is possible to activate the X11 mode.
>    - Intellij is displayed rather smoothly in mode X11 (I didn't notice a
>    slow GUI).
> 
> My 2 cents
> 
> Have fun
> 
> Rudy
> […](#)

And of course, the menus are displayed correctly!

**Comment by rowleya** [comment]
Note if you are on Wayland mode (i.e. just using WSL2 and launching GUI applications without an additional X server), you can "activate X mode" on eclipse by setting the environment variable ```GDK_BACKEND=x11```.  This then works better, but still means that when the window is maximised the menus appear under the start menu.  If you make the window the size it would be when maximised, the menus go over the start menu.  This is how I work with it, avoiding the need to run my own X server.  I guess your solution may be more satisfying though since you get to fully use the maximise buttons... makes me wonder if the Wayland route was the best idea...

**Comment by houdini1968** [comment]
I have passed 'hours' with Gemini to make the IBM ACE toolkit stable...
'Sometimes' when displaying the properties, let's say of a queue listener, *the
gui is frozen.*
The script below, the result of the investigations I did with Gemini.
*Yes at my job for a lot of reasons I am currently stuck with Red Hat 8 :-(*

cat launch-ace.sh
#!/bin/bash

#
==============================================================================
# Script: launch-ace.sh
# Purpose: Safely launch ACE Toolkit with lock cleanup and GUI stability
fixes.
#
==============================================================================

# 1. Environment Configuration
# (ACE_INSTALL_HOME should be exported in your ~/.bashrc)
WORKSPACE_LOCK="/home/m999rvi/IBM/ACET12/onaff-ace/.metadata/.lock"
CONFIG_LOCK_DIR="/home/m999rvi/IBM/ACET12-config/"

# 2. Safety Check: Is the Toolkit already running?
# We use 'pgrep -c' to get a literal 0 if nothing is found (avoids empty
var errors)
# We search for 'javaw-ACE' to avoid accidentally catching the
'IntegrationServer'
GUI_COUNT=$(pgrep -u $USER -c -f "javaw-ACE")

if [ "$GUI_COUNT" -gt 0 ]; then
    echo "------------------------------------------------------------"
    echo "ABORT: ACE Toolkit GUI is already running ($GUI_COUNT processes)."
    echo "Closing this script to prevent workspace corruption."
    echo "------------------------------------------------------------"
    exit 1
fi

# 3. Pre-flight Cleanup
# If we reached here, the GUI is not running. Safe to clear stale locks.
echo "No active Toolkit GUI found. Cleaning up stale locks..."

# Clear the config locks
if [ -d "$CONFIG_LOCK_DIR" ]; then
    find "$CONFIG_LOCK_DIR" -name ".parentlock" -delete 2>/dev/null
fi

# Clear the specific workspace metadata lock you identified
if [ -f "$WORKSPACE_LOCK" ]; then
    rm -f "$WORKSPACE_LOCK"
    echo "CLEANED: $WORKSPACE_LOCK"
fi

# 4. GUI Stability Environment Variables
# Stabilizes the GTK3 event loop for X11/WSL2 and suppresses layout loops

# Core Display Engine
*export GDK_BACKEND=x11*
export SWT_GTK3=1
export GTK_THEME=Adwaita

# Rendering & Memory-to-X11 Protocol Fixes
export GDK_RENDERING=image
export CAIRO_DEBUG_FLUSH=1
export WEBKIT_DISABLE_COMPOSITING_MODE=1

# Event Synchronization & Focus Hardening
# GDK_SYNCHRONIZE prevents race conditions but requires low-noise logs
export GDK_SYNCHRONIZE=1
export GDK_CORE_DEVICE_EVENTS=1
export GTK_CSD=1

*I repeat I use an external X11:*
[image: image.png]


Le mer. 15 avr. 2026 à 10:11, Andrew Rowley ***@***.***> a
écrit :

**Comment by rowleya** [comment]
I just tried instead running eclipse with Vcxsrv.  What is odd is that this worked in exactly the same way - that is, the menu appears behind the windows taskbar at the bottom.  Worse. it isn't possible to make this then overlap the taskbar to solve the problem in the same way.  This implies that the issue is potentially something between how big the eclipse window is, which appears to then not count the extra space for the task bar.

Are you doing something differently with your X server to make this work?

**Comment by cristian-spiescu** [comment]
I recommend X410. It's a commercial X server. The difference to the free one(s)? It has a lot of fixes for quirks, glitches, etc.
We use it for Eclipse w/ success.

**Comment by rowleya** [comment]
Yes, I just tried a trial version to see what happens and this indeed does work better; mostly the menu then appears within the window bounds, but on the odd occasion it "pops out" it then appears over the task bar.  Ideally this would be replicated in WSL itself; I wonder if this is then a fault of the mstsc application that displays the remote desktop view...
